ydb/library/yql/dq/actors/spilling/channel_storage.cpp

```diff
@@ -20,62 +20,109 @@ using namespace NActors;
+        ChannelStorageActor_ = CreateDqChannelStorageActor(txId, channelId, std::move(wakeUp), actorSystem);
+        ChannelStorageActorId_ = ActorSystem_->Register(ChannelStorageActor_->GetActor());
     }
 
     ~TDqChannelStorage() {
-        if (ActorSystem_) {
-            ActorSystem_->Send(
-            new IEventHandle(
-                SelfActorId_,
-                SelfId_,
-                new TEvents::TEvPoison,
-                /*flags=*/0,
-                /*cookie=*/0));
-        } else {
-            TlsActivationContext->AsActorContext().Send(SelfActorId_, new TEvents::TEvPoison);
-        }
+        ActorSystem_->Send(ChannelStorageActorId_, new TEvents::TEvPoison);
     }
 
-    bool IsEmpty() const override {
-        return SelfActor_->IsEmpty();
+    bool IsEmpty() override {
+        UpdateWriteStatus();
+
+        return WritingBlobs_.empty() && StoredBlobsCount_ == 0 && LoadingBlobs_.empty();
     }
 
-    bool IsFull() const override {
-        return SelfActor_->IsFull();
+    bool IsFull() override {
+        UpdateWriteStatus();
+
+        return WritingBlobs_.size() > MAX_INFLIGHT_BLOBS_COUNT || WritingBlobsTotalSize_ > MAX_INFLIGHT_BLOBS_SIZE;
     }
 
     void Put(ui64 blobId, TRope&& blob, ui64 cookie = 0) override {
-        SelfActor_->Put(blobId, std::move(blob), cookie);
+        UpdateWriteStatus();
+
+        auto promise = NThreading::NewPromise<void>();
+        auto future = promise.GetFuture();
+
+        ui64 blobSize = blob.size();
+
+        ActorSystem_->Send(ChannelStorageActorId_, new TEvDqChannelSpilling::TEvPut(blobId, std::move(blob), std::move(promise)), /*flags*/0, cookie);
+
+        WritingBlobs_.emplace(blobId, TWritingBlobInfo(blobSize, std::move(future)));
+        WritingBlobsTotalSize_ += blobSize;
     }
 
     bool Get(ui64 blobId, TBuffer& blob, ui64 cookie = 0) override {
-        return SelfActor_->Get(blobId, blob, cookie);
+        UpdateWriteStatus();
+
+        const auto it = LoadingBlobs_.find(blobId);
+        // If we didn't request loading blob from spilling -> request it
+        if (it == LoadingBlobs_.end()) {
+            auto promise = NThreading::NewPromise<TBuffer>();
+            auto future = promise.GetFuture();
+            ActorSystem_->Send(ChannelStorageActorId_, new TEvDqChannelSpilling::TEvGet(blobId, std::move(promise)), /*flags*/0, cookie);
+
+            LoadingBlobs_.emplace(blobId, std::move(future));
+            return false;
+        }
+        // If we requested loading blob, but it's not loaded -> wait
+        if (!it->second.HasValue()) return false;
+
+        blob = std::move(it->second.ExtractValue());
+        LoadingBlobs_.erase(it);
+        --StoredBlobsCount_;
+
+        return true;
+    }
+
+private:
+    void UpdateWriteStatus() {
+        TVector<THashMap<ui64, TWritingBlobInfo>::iterator> itersToRemove;
+        for (auto it = WritingBlobs_.begin(); it != WritingBlobs_.end();) {
```

Comment:
use erase_if: https://en.cppreference.com/w/cpp/container/unordered_map/erase_if

Reply:
But this is not std::unordered_map, I used THashMap. And THashMap doesn't have erase_if method. Do you suggest to replace it with std::unordered_map?

Reply:
fixed